### ISSUE TITLE
dangerous default value is not detected in case of instantiation (and some other cases)

### ISSUE THREAD

**Issue by mbdevpl** [PyCQA/pylint#788]
Recently I wrote such a program:

```python
"""
    Testing "dangerous default value".
"""
# pylint: disable=too-few-public-methods

class MyFoo:
    """ Foo """
    def __init__(self):
        self.name = 'the answer'

class MyBar:
    """ Bar """
    def __init__(self, my_obj=MyFoo(), my_list=[]): # TWO errrors here!
        super().__init__()
        my_obj.name += ' is'
        self.my_obj = my_obj
        my_list.append('42')
        self.my_list = my_list
    def __str__(self):
        return '\n'.join(['{', '\t' + str(self.my_obj.name), '\t' + str(self.my_list), '}'])

print(MyBar())
print(MyBar())
print(MyBar())
```

Pylint (version 1.6.0 just built from `master`, with astroid 1.5.0 just built from its `master`) notices only one mistake:
```
W: 13, 4: Dangerous default value [] as argument (dangerous-default-value)
```

But as the experiment shows, the side effect of functions being objects also applies to instantiation of `MyFoo`:
```
.../experiments $ python pylint_dangerous_default_value.py
{
        the answer is
        ['42']
}
{
        the answer is is
        ['42', '42']
}
{
        the answer is is is
        ['42', '42', '42']
}

.../experiments $
```

After I fix the error with list:
```python
    def __init__(self, my_obj=MyFoo(), my_list=None): # still ONE errror here!
        super().__init__()
        my_obj.name += ' is'
        self.my_obj = my_obj
        if my_list is None: # dangerous default value fixed
            my_list = []
        my_list.append('42')
        self.my_list = my_list
```

Even if:
```
Global evaluation
-----------------
Your code has been rated at 10.00/10
```

I wouldn't say it's that good:
```
../experiments $ python pylint_dangerous_default_value_fixed.py
{
        the answer is
        ['42']
}
{
        the answer is is
        ['42']
}
{
        the answer is is is
        ['42']
}

.../experiments $
```

Would detection of instantiation in such case be possible to be added to pylint? If yes, where one should look in the code?

**Comment by mbdevpl** [comment]
Another example: default value of type `dict` is detected as dangerous, (causing a message e.g. `Dangerous default value MY_BAZ (builtins.dict) as argument (dangerous-default-value)` but simply inheriting from `dict` lets you bypass the detection. I think any mutable object (complete list here: https://docs.python.org/3/reference/datamodel.html) should be considered a dangerous default value, but currently only instances of selected built-in types are.

```python
"""
    Testing "dangerous default value".
"""
# pylint: disable=too-few-public-methods

class MyFoo:
    """ Foo """
    def __init__(self):
        self.name = 'the answer'

class MyBaz(dict):
    """ Baz """
    pass

class MyBar:
    """ Bar """
    MY_BAZ = MyBaz()
    def __init__(self, my_obj=MyFoo(), my_list=[], my_dict=MY_BAZ): # THREE errrors here!
        super().__init__()
        my_obj.name += ' is' # danger
        self.my_obj = my_obj
        my_list.append('42') # danger
        self.my_list = my_list
        my_dict['answer'] = 42 + my_dict.get('answer', 0) # danger
        self.my_dict = my_dict
    def __str__(self):
        return '\n'.join([
            '{', '\t' + str(self.my_obj.name), '\t' + str(self.my_list), '\t' + str(self.my_dict),
            '}'])

print(MyBar())
print(MyBar())
print(MyBar())
```

```
.../experiments $ python pylint_dangerous_default_values.py
{
        the answer is
        ['42']
        {'answer': 42}
}
{
        the answer is is
        ['42', '42']
        {'answer': 84}
}
{
        the answer is is is
        ['42', '42', '42']
        {'answer': 126}
}

.../experiments $
```

**Comment by textbook** [comment]
On this basis virtually everything would be seen as a dangerous default; even if your custom object is immutable (e.g. inherits a `namedtuple`) it could have mutable attributes.

**Comment by PCManticore** [comment]
I agree with @textbook this is beyond the scope of the dangerous-default-argument, which is used for the most common cases that can trip beginners into Python, such as using as a default a list. Detecting the kind of problem that @mbdevpl describes would require some extra capabilities, such as monitoring the flow of the arguments, which we don't quite have at this moment.

**Comment by mbdevpl** [comment]
Your arguments are convincing. I agree, this might beyond scope of the 
dangerous-default-argument.

Also, I now think that in some cases mutability of default arg might 
even be useful - and not dangerous.

That was enlightening, thanks for the replies!
